fuzzy.go

```diff
@@ -55,35 +54,38 @@ characters upto the first match.
 */
 func Find(pattern string, data []string) []Match {
 	if len(pattern) == 0 {
-		return []Match{}
+		return nil
 	}
-	matches := make([]Match, 0)
+	runes := []rune(pattern)
+	var matches []Match
+	var matchedIndexes []int
 	for i := 0; i < len(data); i++ {
-		candidate := strings.Split(data[i], "")
-		match := Match{
-			MatchedIndexes: make([]int, 0),
+		var match Match
+		match.Str = data[i]
+		match.Index = i
+		if matchedIndexes != nil {
+			match.MatchedIndexes = matchedIndexes
+		} else {
+			match.MatchedIndexes = make([]int, 0, len(pattern))
 		}
 		var score int
 		patternIndex := 0
 		bestScore := -1
 		matchedIndex := -1
 		currAdjacentMatchBonus := 0
-		for j := 0; j < len(candidate); j++ {
-			c := strings.ToLower(candidate[j])
-			p := strings.ToLower(string(pattern[patternIndex]))
-			if p == c {
-				// avoid repeatedly setting params that apply to the whole match
-				if !match.initialized {
-					match.Str = data[i]
-					match.Index = i
-					match.initialized = true
-				}
+		var last rune
+		for j, candidate := range data[i] {
+			if equalFold(candidate, runes[patternIndex]) {
 				score = 0
 				if j == 0 {
 					score += firstCharMatchBonus
 				}
-				score += camelCaseBonus(j, candidate)
-				score += separatorBonus(j, candidate)
+				if unicode.IsLower(last) && unicode.IsUpper(candidate) {
+					score += camelCaseMatchBonus
+				}
+				if j != 0 && isSeparator(last) {
+					score += matchFollowingSeparatorBonus
+				}
 				if len(match.MatchedIndexes) > 0 {
```

Comment:
I don't think this logic works with `j` being a byte iterator rather than a rune iterator. Perhaps keep a separate counter for index? `j` when ranging over a string is the byte index, which breaks adjacency for multi-byte runes. See: https://play.golang.org/p/dVDNZyq6owZ

Reply:
Awesome catch! I guess the problem is that we don't have tests for Unicode.

@ericpauley Could you please fix this and add tests around Unicode handling (especially index tracking and adjacent match bonuses).

Reply:
@lunixbochs Could you explain what the problem is here? Last isn't set using a byte index, it's set in the previous iteration, so multi-length runes should not be an issue. The only place this really needed to be considered was when calculating nextc, where a byte index was needed.

Reply:
I see a few problems with our Unicode handling. Let me add a few review notes for you.

Reply:
Just realized what @lunixbochs meant and made some changes, should be good now as far as I can tell, also added a test case for multi-byte utf-8.